src/network/networkd-dhcp4.c

```diff
@@ -465,6 +465,12 @@ static int dhcp_reset_hostname(Link *link) {
         if (r < 0)
                 return log_link_error_errno(link, r, "DHCP error: Failed to reset transient hostname: %m");
 
+        if (link->network->dhcp_set_static_hostname) {
+                r = manager_set_static_hostname(link->manager, NULL);
+                if (r < 0)
+                        return log_link_error_errno(link, r, "DHCP error: Failed to reset static hostname: %m");
+        }
+
```

Comment:
I do not think we should unset the static hostname. If we do so, then every time we restart or stop networkd, the static hostname is unset. We already unset the transient hostname in the above, so it makes the system's hostname become `FALLBACK_HOSTNAME`, especially every time on shutdown the system...